### PR TITLE
CORTX-33862 Integrate SpLint tool to detect Memory leaks Motr Repo

### PR DESCRIPTION
**Pull request by shailesh-vaidya**
# Problem Statement
- Created Jenkins pipeline to integrate SpLint tool to detect Memory leaks Motr Repo. Jenkins pipeline will take Motr Repo details (Repo URL and Branch) as input and generate a report. The report URL will be sent to the developer as an email notification. 

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM
- [x] Jenkins pipelines used for testing - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/Motr/job/memory-leak-analysis/11/ 

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: jenkins/automation/splint-memory-leak.groovy [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                    STATUS = "UNSTABLE"
+                }
+
+                MemoryLeakAnalysisStatusHTML = "<pre>${clusterStatus}</pre>"
+
+                manager.createSummary("${ICON}").appendText("<h3>Memory Leak Analysis Execution ${currentBuild.currentResult} </h3><p>Please check <a href=\"${BUILD_URL}/artifact/artifacts/splint-analysis.log\">Memory Leak Analysis logs</a> for Memory leak errors</p><p>Please check <a href=\"${BUILD_URL}/console\">Memory Leak Analysis Execution logs</a> for more info</p>", false, false, false, "red")
+
+                // Email Notification
+                env.build_stage = "${build_stage}"
+                env.cluster_status = "${MemoryLeakAnalysisStatusHTML}"
```
Comment (gauravchaudhari02) [inline jenkins/automation/splint-memory-leak.groovy:78]: can you change variable name to appropriate name on what it holds?

File: jenkins/automation/splint-memory-leak.groovy [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                script {
+                    // Archive Deployment artifacts in jenkins build
+                    archiveArtifacts artifacts: "artifacts/*.*", onlyIfSuccessful: false, allowEmptyArchive: true
+                }
+        }
+
+        always {
+            script {
+                // Jenkins Summary
+                clusterStatus = ""
```
Comment (gauravchaudhari02) [inline jenkins/automation/splint-memory-leak.groovy:54]: do we need this?

File: jenkins/automation/splint-memory-leak.groovy [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+                // Email Notification
+                env.build_stage = "${build_stage}"
+                env.cluster_status = "${MemoryLeakAnalysisStatusHTML}"
+                def recipientProvidersClass = [[$class: 'RequesterRecipientProvider']]
+                mailRecipients = "CORTX.DevOps.RE@seagate.com"
+                emailext (
+                    body: '''${SCRIPT, template="devops-automation-email.template"}''',
+                    mimeType: 'text/html',
+                    subject: "[Jenkins Build ${currentBuild.currentResult}] : ${env.JOB_NAME}",
```
Comment (gauravchaudhari02) [inline jenkins/automation/splint-memory-leak.groovy:84]: can you add `MESSAGE` as a subject?

File: jenkins/automation/splint-memory-leak.groovy (hunk 1 above)
Comment (shailesh-vaidya) [reply jenkins/automation/splint-memory-leak.groovy:78]: We will store metrics collected by parsing raw splint tool logs in MemoryLeakAnalysisStatusHTML variable

File: jenkins/automation/splint-memory-leak.groovy (hunk 3 above)
Comment (shailesh-vaidya) [reply jenkins/automation/splint-memory-leak.groovy:84]: What `MESSAGE` need to be added? Can you elaborate more?

File: jenkins/automation/splint-memory-leak.groovy (hunk 2 above)
Comment (shailesh-vaidya) [reply jenkins/automation/splint-memory-leak.groovy]: Renamed variable name to `MemoryLeakStats`

File: jenkins/automation/splint-memory-leak.groovy (hunk 3 above)
Comment (shailesh-vaidya) [reply jenkins/automation/splint-memory-leak.groovy]: Fixed. Please review again.

File: jenkins/automation/splint-memory-leak.groovy [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            }
+        }
+
+        stage ('Execute Memory Analysis') {
+            steps {
+                script { build_stage = env.STAGE_NAME }
+                sh label: '', returnStatus: true, script: '''
+                yum install splint -y
+                mkdir -p artifacts
+                for file in  $(find $PWD -type f -regextype posix-egrep -regex ".+\\\\\\\\.(c|h|lcl)$"); do splint +trytorecover -preproc -warnposix +line-len 100000 $file | sed \'H;1h;$!d;g;s/\\n  */ /g\' >> artifacts/splint-analysis.log ; done || echo "Log Generated"
```
Comment (gauravchaudhari02) [inline jenkins/automation/splint-memory-leak.groovy:36]: @shailesh-vaidya, can we echo message to `ERROR: splint execution failed. please check logs for more details`

File: jenkins/automation/splint-memory-leak.groovy (hunk 1 above)
Comment (gauravchaudhari02) [reply jenkins/automation/splint-memory-leak.groovy:78]: @shailesh-vaidya, can we change variable `env.cluster_status` as well to something like `env.memory_leak_stats`

Comment (shailesh-vaidya) [reply jenkins/automation/splint-memory-leak.groovy]: Fixed and tested in https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/Motr/job/memory-leak-analysis/19/

File: jenkins/automation/splint-memory-leak.groovy (hunk 4 above)
Comment (shailesh-vaidya) [reply jenkins/automation/splint-memory-leak.groovy]: Fixed and tested in https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/Motr/job/memory-leak-analysis/19/

**Comment by shailesh-vaidya** [comment]
> https://github.com/Seagate/cortx-re/pull/1169#discussion_r942183380

@gauravchaudhari02  - Fixed and tested in https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/Motr/job/memory-leak-analysis/19/

**Review by gauravchaudhari02 — APPROVED** [review]

**Pull request merged**